src/NServiceBus.Core.Tests/Recoverability/FailureInfoStorageTests.cs

```diff
@@ -0,0 +1,119 @@
+
+            var storage = GetFailureInfoStorage();
+
+            storage.RecordFailedProcessingAttempt(messageId, exceptionDispatchInfo);
+            
+            var failureInfo = storage.GetFailureInfoForMessage(messageId);
+
+            Assert.NotNull(failureInfo);
+            Assert.AreEqual(1, failureInfo.FailedProcessingAttempts);
+            Assert.AreSame(exception, failureInfo.Exception);
+        }
+
+        [Test]
+        public void When_recording_failure_many_times_should_store_number_of_attempts_and_last_exception()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+            var secondException = new Exception();
+
+            var storage = GetFailureInfoStorage();
+
+            storage.RecordFailedProcessingAttempt(messageId, ExceptionDispatchInfo.Capture(new Exception()));
+            storage.RecordFailedProcessingAttempt(messageId, ExceptionDispatchInfo.Capture(secondException));
+
+            var failureInfo = storage.GetFailureInfoForMessage(messageId);
+
+            Assert.NotNull(failureInfo);
+            Assert.AreEqual(2, failureInfo.FailedProcessingAttempts);
+            Assert.AreSame(secondException, failureInfo.Exception);
+        }
+
+        [Test]
+        public void When_clearing_failure_should_return_null_on_subsequent_retrieval()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+
+            var storage = GetFailureInfoStorage();
+
+            storage.RecordFailedProcessingAttempt(messageId, ExceptionDispatchInfo.Capture(new Exception()));
+            
+            var failureInfo = storage.GetFailureInfoForMessage(messageId);
+            Assert.NotNull(failureInfo);
+            
+            storage.ClearFailureInfoForMessage(messageId);
+
+            failureInfo = storage.GetFailureInfoForMessage(messageId);
+            Assert.AreSame(ProcessingFailureInfo.NullFailureInfo, failureInfo);
+        }
+
+        [Test]
+        public void When_recording_more_than_max_number_of_failures_should_remove_least_recently_used_entry()
+        {
+            const int MaxElements = 50;
+            var storage = new FailureInfoStorage(maxElements: MaxElements);
+
+            var lruMessageId = Guid.NewGuid().ToString("D");
+
+            storage.RecordFailedProcessingAttempt(lruMessageId, ExceptionDispatchInfo.Capture(new Exception()));
+
+            for (var i = 0; i < MaxElements; ++i)
+            {
+                var messageId = Guid.NewGuid().ToString("D");
+                var exception = new Exception();
+
+                storage.RecordFailedProcessingAttempt(messageId, ExceptionDispatchInfo.Capture(exception));
+            }
+
+            var lruFailureInfo = storage.GetFailureInfoForMessage(lruMessageId);
+            Assert.AreSame(ProcessingFailureInfo.NullFailureInfo, lruFailureInfo);
+        }
+
+        [Test]
+        public void When_recording_more_than_max_number_of_failures_many_times_should_remove_least_recently_used_entry()
```

Comment:
This doesn't test what it's name implies. It proves that an item which is not "least recently used" is not removed from the collection.